skfda/misc/metrics/_sdtw_distances.py

```diff
@@ -0,0 +1,368 @@
+        raise ValueError(
+            "The cost function does not return a symmetric matrix"
+        )
+
+
+def _check_input_fdata(
+    fdata1,
+    fdata2
+) -> None:
+    """check compatible, one sample and one-dimensional"""
+    _check_compatible_fdata(fdata1, fdata2)
+
+    if not isinstance(fdata1, FDataGrid):
+        raise ValueError(
+            "fdata1 and fdata2 must be FDataGrid objects"
+        )
+
+    # dim_codomain can be higher than one,
+    # but dim_domain must be one
+    if not fdata1.dim_domain == 1:
+        raise ValueError(
+            "fdata1 and fdata2 must have their domain dimension equal to one"
+        )
+
+    if not (fdata1.n_samples == 1 and fdata2.n_samples == 1):
+        raise ValueError(
+            "Both fdata1 and fdata2 must each contain a single sample"
+        )
+
+
+def half_sq_euclidean(
+    arg1: NDArrayFloat,
+    arg2: Optional[NDArrayFloat] = None,
+) -> NDArrayFloat:
+    """
+    Return the half squared Euclidean row distance between
+    two matrices of size (n1, d) and (n2, d).
+
+    Args:
+        arg1: First sample (n_samples_1, n_features).
+        arg2: Second sample (n_samples_2, n_features).
+            If it is ``None`` returns the half squared Euclidean distance
+            between the rows of ``arg1``.
+
+    The distance is taken w.r.t matrix rows.
+    """
+    if arg2 is not None:
+        # sq_euclidean(X, Y) =
+        # DotProduct(sigma_0_bounds=(10**-10, 10**10)) =
+        # diag(np.dot(X, X.T)) + diag(np.dot(Y, Y.T)) - 2*np.dot(X, Y.T)
+        # np.dot(X, Y.T)=inner_product_matrix(X, Y)
+        cost_12 = -1 * inner_product_matrix(arg1, arg2)
+
+        # np.sum(X ** 2, axis=1)=inner_product(X, X)
+        cost_12 += (0.5 * inner_product(arg1, arg1))[:, np.newaxis]
+        cost_12 += 0.5 * inner_product(arg2, arg2)
+
+        return cost_12
+    else:
+        cost_11 = -1 * inner_product_matrix(arg1, arg2)
+        # half Sum Of Squares on each row
+        sos_row = 0.5 * inner_product(arg1, arg1)
+        cost_11 += sos_row[:, np.newaxis]
+        cost_11 += sos_row
+
+        return cost_11
+
+
+def _sdtw_divergence(
+    fdata1: FDataGrid,
+    fdata2: FDataGrid,
+    *,
+    gamma: float = 1.0,
+    cost: Union[Tuple[NDArrayFloat], Callable[[
+        NDArrayFloat, NDArrayFloat, NDArrayFloat]]] = None,
+    check_cost: bool = False,
+    **kwargs_cost: Dict
+) -> float:
+
+    # fdata1 and fdata2 are only required to be one-dimensional.
+    # They can have different (one-dimensional) domain ranges and
+    # different grid points size
+
+    _check_input_fdata(fdata1, fdata2)
+
+    n1, n2 = len(fdata1.grid_points[0]), len(fdata2.grid_points[0])
+
+    if gamma <= 0:
+        raise ValueError(
+            f"gamma was set to {gamma} but must be positive"
+        )
+
+    # booleans to remember the input format of cost if given
+    cost_was_tuple = isinstance(cost, tuple)
+    cost_was_callable = isinstance(cost, Callable)
```

Comment:
For this is better to use the Python `callable` function, that checks if something is callable.